Networking/Networking/Mapper/ProductTagListMapper.swift

```diff
@@ -0,0 +1,64 @@
+import Foundation
+
+/// Mapper: ProductTag List
+///
+struct ProductTagListMapper: Mapper {
+    /// Site Identifier associated to the `ProductTags`s that will be parsed.
+    ///
+    /// We're injecting this field via `JSONDecoder.userInfo` because SiteID is not returned in any of the ProductTag Endpoints.
+    ///
+    let siteID: Int64
+
+    /// (Attempts) to convert a dictionary into [ProductTag].
+    ///
+    func map(response: Data) throws -> [ProductTag] {
+        let decoder = JSONDecoder()
+        decoder.userInfo = [
+            .siteID: siteID
+        ]
+
+        let decodedResponse = try? decoder.decode(ProductTagListEnvelope.self, from: response)
+        let decodedResponseBatchUpdatedTags = try? decoder.decode(ProductTagListBatchUpdateEnvelope.self, from: response)
+
+        return decodedResponse?.tags ??
+            decodedResponseBatchUpdatedTags?.createdTags ??
+            decodedResponseBatchUpdatedTags?.deletedTags ?? []
```

Comment:
> it looks like the major difference between `ProductTagListEnvelope` and `ProductTagListBatchUpdateEnvelope` is the key that maps to the tags data

In reality, this is not the only difference, because `create` and `delete` are nested under `data`.

`ProductTagListEnvelope` is used to parse a JSON like that, where `data` is the only key (like all the others endpoints):
```
{
    "data": [
      {
        "id": 34,
        "name": "Leather Shoes",
        "slug": "leather-shoes",
        "description": "",
        "count": 0,
        "_links": {
          "self": [
            {
              "href": "https://example.com/wp-json/wc/v3/products/tags/34"
            }
          ],
          "collection": [
            {
              "href": "https://example.com/wp-json/wc/v3/products/tags"
            }
          ]
        }
      }
}
```

`ProductTagListBatchUpdateEnvelope` is a little bit different, and it's the first time that we need to parse a case like that. Basically, the result are nested keys under `data`, so passing an enum of `CodingKeys` it's not really helpful.
This is a JSON example:
```
{
    "data": {
      "create": [
        {
          "id": 36,
          "name": "Round toe",
          "slug": "round-toe",
          "description": "",
          "count": 0,
          "_links": {
            "self": [
              {
                "href": "https://example.com/wp-json/wc/v3/products/tags/36"
              }
            ],
            "collection": [
              {
                "href": "https://example.com/wp-json/wc/v3/products/tags"
              }
            ]
          }
        }
      ]
    }
}

```

Reply:
thanks for the JSON examples, they were helpful for understanding the response structure!

I just felt that it could be more efficient and explicit if we don't have try decoding with both types of envelope for all types of responses, by passing in an enum for the type of response the caller expects. something like:

```swift
enum ProductTagListMapper.ResponseType {
  case load
  case create
  case delete
}
```

so that `ProductTagListMapper` can be initialized as `ProductTagListMapper(siteID:responseType:)` in `ProductTagsRemote`.

then we have three types of envelope: the existing `ProductTagListEnvelope` for `load` case, and splitting `ProductTagListBatchUpdateEnvelope` into two - `ProductTagListBatchCreateEnvelope` for `create` case and `ProductTagListBatchDeleteEnvelope` for `delete` case.

in this `func map(response: Data) throws -> [ProductTag]`, we can now return the response based on the case:

```swift
switch responseType {
  case .load:
    return try decoder.decode(ProductTagListEnvelope.self, from: response).tags
  case .create:
    return try decoder.decode(ProductTagListBatchCreateEnvelope.self, from: response).tags
  case .delete:
    return try decoder.decode(ProductTagListBatchDeleteEnvelope.self, from: response).tags
}
```

what do you think?

Reply:
Sounds good to me! Updated here 2f8851766fd2539fbf9ed23c0e23cd0ddfb3479f and here 9711995cb027cea06724df3df32ae5763839790a